daffodil-lib/src/main/scala/org/apache/daffodil/xml/PUARemappers.scala

```diff
@@ -0,0 +1,117 @@
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.daffodil.xml
+
+import org.apache.daffodil.exceptions.Assert
+import org.apache.daffodil.util.CharacterSetRemapper
+
+object Unicode {
+  def isLeadingSurrogate(c: Int) = {
+    c >= 0xD800 && c <= 0xDBFF
+  }
+
+  def isTrailingSurrogate(c: Int) = {
+    c >= 0xDC00 && c <= 0xDFFF
+  }
+}
+/**
+ * Remaps illegal XML chars to the PUA, and optionally CR also to the PUA.
+ *
+ * Handles unpaired unicode surrogate code points properly (remaps them).
+ *
+ * Legal XML v1.0 chars are #x9 | #xA | #xD | [#x20-#xD7FF] | [#xE000-#xFFFD] | [#x10000-#x10FFFF]
+ *
+ * Normally XML also remaps CRLF to LF and CR (isolated) to LF, but this is problematic
+ * when data must be preserved perfectly so we have options to turn that off.
+ *
+ * We also can check and error if the string contains conflicting PUA characters to begin with.
+ */
+final class RemapXMLIllegalCharToPUA (
+  checkForExistingPUA: Boolean,
+  replaceCRWithLF: Boolean)
+  extends CharacterSetRemapper {
+  import Unicode._
+
+  /**
+   * Remaps to PUA. Note return is negated char code of replacement char if we're to skip a character
+   */
+  override protected def remap(prev: Int, curr: Int, next: Int): Int = {
+    val res: Int = curr match {
+      case 0x9 => curr
+      case 0xA => curr
+      case 0xD =>
+        if (next == 0xA) {
+          // CRLF case.
+          if (replaceCRWithLF)
+            -0xA // CRLF => LF, standard XML behavior. Note negated.
+          else
+            0xE00D // remap CR to preserve it. Leave LF alone.
+        } else {
+          // isolated CR case
+          if (replaceCRWithLF)
+            0xA // isolated CR => LF, standard XML behavior. Note NOT negated.
+          else
+            0xE00D // remap isolated CR to LF.
+        }
+      case _ if (curr < 0x20) => curr + 0xE000 // ascii c0 controls
+      case _ if (curr > 0xD7FF && curr < 0xE000) => { // surrogate chars
+        if ((isLeadingSurrogate(curr) && isTrailingSurrogate(next)) ||
+          (isTrailingSurrogate(curr) && isLeadingSurrogate(prev))) {
+          // well formed surrogate pairs are preserved
+          curr
+        } else {
+          // curr is an isolated surrogate, so to preserve we must remap to PUA
+          curr + 0x1000
+        }
+      }
+      case _ if (curr >= 0xE000 && curr <= 0xF8FF) => { // Unicode PUA is E000 to F8FF.
+        if (checkForExistingPUA)
+          Assert.usageError("Pre-existing Private Use Area (PUA) character found in data: '%s'".format(curr))
```

Comment:
Should this throw an actual exception that can be handled and something done with? If a caller has enabled the `checkForExistingPUA` flag might they want to convert that to an SDE or something?